### PR TITLE
feat: add useERC20TokensDetailed

### PR DESCRIPTION
**Pull request by zhouhanseng**
closes https://github.com/DimensionDev/Maskbook/issues/3605

### CODE REVIEW

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts [hunk 1]
```diff
@@ -1,66 +1,107 @@
...
-        asyncTokenSymbol,
-        asyncTokenSymbolBytes32,
-        asyncTokenDecimals,
-    ]
+async function getToken(
+    erc20TokenContract: ERC20 | null,
+    erc20TokenBytes32Contract: ERC20Bytes32 | null,
+    token: Partial<ERC20TokenDetailed> | undefined,
+) {
+    const name = token?.name ?? (await (erc20TokenContract?.methods.name().call() ?? ''))
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:80]: Refactor this function with `Promise.allSettled` to get benefits:

+ To batch-request many promises.
+ To recover from the single point of failure.

File: packages/web3-shared/src/hooks/useNativeTokenDetailed.ts [hunk 2]
```diff
@@ -1,21 +1,45 @@
...
-            chainId,
-            name: nativeCurrency?.name ?? 'Unknown',
-            symbol: nativeCurrency?.symbol ?? '',
-            decimals: nativeCurrency?.decimals ?? 0,
-        }
+        return listOfToken.map(() => createTokenDetailed(NATIVE_TOKEN_ADDRESS, chainId, nativeCurrency))
     }, [chainId, NATIVE_TOKEN_ADDRESS])
 }
+
+function createTokenDetailed(
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useNativeTokenDetailed.ts:26]: Use `createNativeToken` instead.

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 1 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:109]: done

File: packages/web3-shared/src/hooks/useNativeTokenDetailed.ts (hunk 2 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useNativeTokenDetailed.ts]: done

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts [hunk 3]
```diff
@@ -1,66 +1,115 @@
...
+import type { ERC20Bytes32 } from '@masknet/contracts/types/ERC20Bytes32'
+import { useERC20TokenContract, useERC20TokenContracts } from '../contracts/useERC20TokenContract'
+import { useERC20TokenBytes32Contract, useERC20TokenBytes32Contracts } from '../contracts/useERC20TokenBytes32Contract'
+import { parseStringOrBytes32, createLoadingFailToken, createERC20Token } from '../utils'
+
+export const LOADING_FAIL_TOKEN_NAME = 'TOKEN LOADING FAIL'
+
+export function checkTokenLoadingFailed(name: string) {
+    return name === LOADING_FAIL_TOKEN_NAME
+}
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:16]: They are not using.

File: packages/web3-shared/src/hooks/useTokenDetailed.ts [hunk 4]
```diff
@@ -34,3 +44,34 @@ export function useTokenDetailed<P extends EthereumTokenType, Q extends Ethereum
             unreachable(type_)
     }
 }
+
+export function useFungibleTokensDetailed(listOfToken: Pick<FungibleToken, 'address' | 'type'>[]) {
+    const nativeTokenOrders: number[] = []
+    const listOfNativeToken = listOfToken.filter((t, i) => {
+        if (t.type === EthereumTokenType.Native) {
+            nativeTokenOrders.push(i)
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useTokenDetailed.ts:52]: Don't make side-effect in a filter.

File: packages/web3-shared/src/hooks/useTokenDetailed.ts [hunk 5]
```diff
@@ -34,3 +44,34 @@ export function useTokenDetailed<P extends EthereumTokenType, Q extends Ethereum
...
+    )
+    const { value: erc20TokensDetailed = [], ...asyncErc20Result } = useERC20TokensDetailed(
+        listOfERC20Token as Pick<ERC20Token, 'type' | 'address'>[],
+    )
+
+    const asyncList = [asyncNativeResult, asyncErc20Result]
+
+    const tokensDetailed = erc20TokensDetailed
+
+    nativeTokenOrders.forEach((order, i) => tokensDetailed.splice(order, 0, nativeTokensDetailed[i]))
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useTokenDetailed.ts:70]: It looks like this variable can be obliterated.

File: packages/web3-shared/src/hooks/useNativeTokenDetailed.ts [hunk 6]
```diff
@@ -1,21 +1,18 @@
...
+
+    return useAsyncRetry(async (): Promise<NativeTokenDetailed> => createNativeToken(chainId), [chainId])
+}
+
+export function useNativeTokensDetailed(listOfToken: Pick<NativeToken, 'address' | 'type'>[]) {
+    const chainId = useChainId()
+
+    return useAsyncRetry(async (): Promise<FungibleTokenDetailed[]> => {
+        return listOfToken.map(() => createNativeToken(chainId))
+    }, [chainId])
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useNativeTokenDetailed.ts:17]: Do you forget to add the `listOfToken` into the dependence list?

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts [hunk 7]
```diff
@@ -1,66 +1,115 @@
...
+                const erc20TokenContract = erc20TokenContracts[i]
+                const erc20TokenBytes32Contract = erc20TokenBytes32Contracts[i]
+                const _token = await getToken(erc20TokenContract, erc20TokenBytes32Contract, token)
+                return { address: token.address, ..._token }
+            }),
+        )
+    }, [
+        chainId,
+        JSON.stringify(erc20TokenContracts),
+        JSON.stringify(erc20TokenBytes32Contracts),
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:73]:
```suggestion
        JSON.stringify(listOfAddress),
```

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts [hunk 8]
```diff
@@ -1,66 +1,115 @@
...
+                          ? createERC20Token(
+                                chainId,
+                                token.value.address,
+                                typeof token.value.decimals === 'string'
+                                    ? Number(token.value.decimals)
+                                    : token.value.decimals,
+                                parseStringOrBytes32(token.value.name, token.value.nameBytes32, 'Unknown Token'),
+                                parseStringOrBytes32(token.value.symbol, token.value.symbolBytes32, 'Unknown'),
+                            )
+                          : createLoadingFailToken(listOfToken[i].address, chainId, token.reason),
```
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:91]: Why do we need loading to fail token here? It's not a token at all.

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 1 above)
Comment (guanbinrui) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:109]: It's still a sequential promise invoking.

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 3 above)
Comment (septs) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:16]:
```suggestion
export const LOADING_FAIL_TOKEN_NAME = 'TOKEN LOADING FAIL'

export function isTokenLoadingFailed(name: string) {
    return name === LOADING_FAIL_TOKEN_NAME
}
```

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 8 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:91]: Save the address of fail token to retry it again.

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 1 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:109]: done

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 3 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:16]: removed

File: packages/web3-shared/src/hooks/useTokenDetailed.ts (hunk 4 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useTokenDetailed.ts:52]: done

File: packages/web3-shared/src/hooks/useTokenDetailed.ts (hunk 5 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useTokenDetailed.ts:70]: removed

File: packages/web3-shared/src/hooks/useNativeTokenDetailed.ts (hunk 6 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useNativeTokenDetailed.ts:17]: done

File: packages/web3-shared/src/hooks/useERC20TokenDetailed.ts (hunk 7 above)
Comment (zhouhanseng) [reply packages/web3-shared/src/hooks/useERC20TokenDetailed.ts:73]: done

**Review by septs — APPROVED** [review]

**Pull request merged**